src/render/draw_heatmap.js

```diff
@@ -0,0 +1,149 @@
+// @flow
+
+const mat4 = require('@mapbox/gl-matrix').mat4;
+const Texture = require('./texture');
+const pixelsToTileUnits = require('../source/pixels_to_tile_units');
+
+import type Painter from './painter';
+import type SourceCache from '../source/source_cache';
+import type HeatmapStyleLayer from '../style/style_layer/heatmap_style_layer';
+import type HeatmapBucket from '../data/bucket/heatmap_bucket';
+import type TileCoord from '../source/tile_coord';
+
+module.exports = drawHeatmap;
+
+function drawHeatmap(painter: Painter, sourceCache: SourceCache, layer: HeatmapStyleLayer, coords: Array<TileCoord>) {
+    if (painter.isOpaquePass) return;
+
+    const gl = painter.gl;
+
+    painter.setDepthSublayer(0);
+    painter.depthMask(false);
+
+    // Allow kernels to be drawn across boundaries, so that
+    // large kernels are not clipped to tiles
+    gl.disable(gl.STENCIL_TEST);
+
+    renderToTexture(gl, painter, layer);
+
+    gl.clearColor(0, 0, 0, 0);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+
+    // Turn on additive blending for kernels, which is a key aspect of kernel density estimation formula
+    gl.blendFunc(gl.ONE, gl.ONE);
+
+    for (let i = 0; i < coords.length; i++) {
+        const coord = coords[i];
+
+        // Skip tiles that have uncovered parents to avoid flickering; we don't need
+        // to use complex tile masking here because the change between zoom levels is subtle,
+        // so it's fine to simply render the parent until all its 4 children are loaded
+        if (sourceCache.hasRenderableParent(coord)) continue;
+
+        const tile = sourceCache.getTile(coord);
+        const bucket: ?HeatmapBucket = (tile.getBucket(layer): any);
+        if (!bucket) continue;
+
+        const programConfiguration = bucket.programConfigurations.get(layer.id);
+        const program = painter.useProgram('heatmap', programConfiguration);
+        const {zoom} = painter.transform;
+        programConfiguration.setUniforms(gl, program, layer, {zoom});
+        gl.uniform1f(program.uniforms.u_radius, layer.getPaintValue('heatmap-radius', {zoom}));
+
+        gl.uniform1f(program.uniforms.u_extrude_scale, pixelsToTileUnits(tile, 1, zoom));
+
+        gl.uniform1f(program.uniforms.u_intensity, layer.getPaintValue('heatmap-intensity', {zoom}));
+        gl.uniformMatrix4fv(program.uniforms.u_matrix, false, coord.posMatrix);
+
+        program.draw(
+            gl,
+            gl.TRIANGLES,
+            layer.id,
+            bucket.layoutVertexBuffer,
+            bucket.indexBuffer,
+            bucket.segments,
+            programConfiguration);
+    }
+
+    renderTextureToMap(gl, painter, layer);
+}
+
+function renderToTexture(gl, painter, layer) {
+    gl.activeTexture(gl.TEXTURE1);
+
+    // Use a 4x downscaled screen texture for better performance
+    gl.viewport(0, 0, painter.width / 4, painter.height / 4);
```

Comment:
Do we need to account for the device pixel ratio here? E.g. do we need a "retina" version of the downscaled texture?

Reply:
I think not because `painter.width` and `painter.height` already have "retina" values.